{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5925d2de-0cba-4544-8e21-d897b83b0f15",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home1/zhuyixin/.conda/envs/assembly/lib/python3.10/site-packages/fpdf/__init__.py:39: UserWarning: You have both PyFPDF & fpdf2 installed. Both packages cannot be installed at the same time as they share the same module namespace. To only keep fpdf2, run: pip uninstall --yes pypdf && pip install --upgrade fpdf2\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import sys\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import argparse\n",
    "import os\n",
    "import warnings\n",
    "import pandas as pd\n",
    "from fpdf import FPDF\n",
    "import cairosvg\n",
    "from IPython.display import SVG, display\n",
    "import math\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5d2eb9c5-77cc-4861-9288-8ae8756172d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "species = \"mPonAbe1\"\n",
    "haploid = False\n",
    "gene = \"IGH\"\n",
    "AssType = \"hap2\"\n",
    "dirOut = f\"/home1/zhuyixin/zhuyixin_proj/AssmQuality/errorPlots/{species}\"\n",
    "dirStat = f\"/home1/zhuyixin/zhuyixin_proj/AssmQuality/errorStats/{species}\"\n",
    "bam_file_path = f\"/home1/zhuyixin/zhuyixin_proj/AssmQuality/aligned_bam/{species}/{species}_merged_sorted_primary.bam\"\n",
    "if (AssType == \"mat\") or (AssType == \"hap1\"):\n",
    "    atype = \"pri\"\n",
    "else:\n",
    "    atype = \"alt\"\n",
    "pri_pileup_file = f'../errorStats/{species}/{gene}_{atype}_pileup.txt'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f114ec98-14c5-4428-a553-b9b6920c243c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_read_bases(read_bases):\n",
    "    \"\"\"\n",
    "    Parse read bases to count correct matches (., and ,) while correctly handling indels,\n",
    "    read starts (^), read ends ($), and skipping over sequences indicating indels (+nXXX or -nXXX).\n",
    "    \"\"\"\n",
    "    cleaned_bases = re.sub(r'\\^.', '', read_bases)\n",
    "    cleaned_bases = cleaned_bases.replace('$', '')\n",
    "    cleaned_bases = re.sub(r'[\\+\\-](\\d+)([ACGTNacgtn]+)', '', cleaned_bases)\n",
    "    cleaned_bases = cleaned_bases.replace('*', '')\n",
    "    correct = cleaned_bases.count('.') + cleaned_bases.count(',')\n",
    "\n",
    "    return correct\n",
    "\n",
    "\n",
    "import re\n",
    "\n",
    "def count_indels(read_bases):\n",
    "    \"\"\"\n",
    "    Count the number of indels in the read bases, handling both insertions and deletions.\n",
    "    Indels are indicated by + or - followed by the length of the indel and the sequence.\n",
    "    \"\"\"\n",
    "\n",
    "    # Find all occurrences of indels, both insertions and deletions\n",
    "    indels = re.findall(r'[\\+\\-](\\d+)([ACGTNacgtn]+)', read_bases)\n",
    "\n",
    "    # The total count of indels is the sum of the occurrences\n",
    "    indel_count = sum(int(length) for length, _ in indels)\n",
    "\n",
    "    return indel_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "42cf60b6-a561-4ca4-80f4-7aa77d690bc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pos</th>\n",
       "      <th>Correct</th>\n",
       "      <th>PercentCorrect</th>\n",
       "      <th>Depth</th>\n",
       "      <th>Indel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.215789e+06</td>\n",
       "      <td>1.215789e+06</td>\n",
       "      <td>1.215789e+06</td>\n",
       "      <td>1.215789e+06</td>\n",
       "      <td>1.215789e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1.031147e+08</td>\n",
       "      <td>4.822201e+01</td>\n",
       "      <td>9.989388e+01</td>\n",
       "      <td>4.827318e+01</td>\n",
       "      <td>1.117924e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>3.509682e+05</td>\n",
       "      <td>8.167864e+00</td>\n",
       "      <td>5.941782e-01</td>\n",
       "      <td>8.171534e+00</td>\n",
       "      <td>1.989873e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.025068e+08</td>\n",
       "      <td>2.200000e+01</td>\n",
       "      <td>6.111111e+01</td>\n",
       "      <td>2.400000e+01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.028107e+08</td>\n",
       "      <td>4.300000e+01</td>\n",
       "      <td>1.000000e+02</td>\n",
       "      <td>4.300000e+01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.031147e+08</td>\n",
       "      <td>4.800000e+01</td>\n",
       "      <td>1.000000e+02</td>\n",
       "      <td>4.800000e+01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.034186e+08</td>\n",
       "      <td>5.400000e+01</td>\n",
       "      <td>1.000000e+02</td>\n",
       "      <td>5.400000e+01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.037226e+08</td>\n",
       "      <td>7.500000e+01</td>\n",
       "      <td>1.000000e+02</td>\n",
       "      <td>7.500000e+01</td>\n",
       "      <td>1.861000e+03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Pos       Correct  PercentCorrect         Depth         Indel\n",
       "count  1.215789e+06  1.215789e+06    1.215789e+06  1.215789e+06  1.215789e+06\n",
       "mean   1.031147e+08  4.822201e+01    9.989388e+01  4.827318e+01  1.117924e-01\n",
       "std    3.509682e+05  8.167864e+00    5.941782e-01  8.171534e+00  1.989873e+00\n",
       "min    1.025068e+08  2.200000e+01    6.111111e+01  2.400000e+01  0.000000e+00\n",
       "25%    1.028107e+08  4.300000e+01    1.000000e+02  4.300000e+01  0.000000e+00\n",
       "50%    1.031147e+08  4.800000e+01    1.000000e+02  4.800000e+01  0.000000e+00\n",
       "75%    1.034186e+08  5.400000e+01    1.000000e+02  5.400000e+01  0.000000e+00\n",
       "max    1.037226e+08  7.500000e+01    1.000000e+02  7.500000e+01  1.861000e+03"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_pri = {}  # To store results, keyed by (chromosome, position)\n",
    "with open(pri_pileup_file, 'r') as f:\n",
    "    for line in f:\n",
    "        chrom, pos, ref_base, depth, read_bases, _ = line.split()[:6]\n",
    "        correct = parse_read_bases(read_bases)\n",
    "        indel = count_indels(read_bases)            \n",
    "        if int(depth) > 0:\n",
    "            percent_correct = (correct / int(depth)) * 100\n",
    "        else:\n",
    "            percent_correct = 0\n",
    "         # Store or process results\n",
    "        results_pri[(chrom, pos)] = (correct, percent_correct, depth, indel)\n",
    "\n",
    "pri_p = [(*key, *value) for key, value in results_pri.items()]\n",
    "pri_pileup = pd.DataFrame(pri_p, columns=['Chrom', 'Pos', 'Correct', 'PercentCorrect', \"Depth\", \"Indel\"])\n",
    "pri_pileup['Pos'] = pd.to_numeric(pri_pileup['Pos'])\n",
    "pri_pileup['Correct'] = pd.to_numeric(pri_pileup['Correct'])\n",
    "pri_pileup['PercentCorrect'] = pd.to_numeric(pri_pileup['PercentCorrect'])\n",
    "pri_pileup['Depth'] = pd.to_numeric(pri_pileup['Depth'])\n",
    "pri_pileup['Indel'] = pd.to_numeric(pri_pileup['Indel'])\n",
    "pri_pileup.sort_values(by=['Chrom', 'Pos'], inplace=True)\n",
    "pri_pileup.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "a1dbdb63-7329-4374-b493-5448e2b85231",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9590159147681053"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pri_pileup[pri_pileup['PercentCorrect']> 99].shape[0]/pri_pileup.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "df09d9ca-0e5e-4ec0-b0e3-e8cb955f6a17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>read_length</th>\n",
       "      <th>mapping_quality</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>mismatch_rate</th>\n",
       "      <th>longindels</th>\n",
       "      <th>total_indel_length</th>\n",
       "      <th>indel_rate</th>\n",
       "      <th>soft_clipping</th>\n",
       "      <th>hard_clipping</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3866</th>\n",
       "      <td>m54329U_220313_173323/117639354/ccs</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102481413</td>\n",
       "      <td>28540</td>\n",
       "      <td>30</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>102509952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3867</th>\n",
       "      <td>m54329U_210404_020346/15336368/ccs</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102487172</td>\n",
       "      <td>19795</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>102506966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3868</th>\n",
       "      <td>m54329U_220313_173323/76744132/ccs</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102487651</td>\n",
       "      <td>28758</td>\n",
       "      <td>60</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>102516408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3869</th>\n",
       "      <td>m64076_220312_125202/28902388/ccs</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102487840</td>\n",
       "      <td>21075</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>102508914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3870</th>\n",
       "      <td>m64076_220313_224245/157485750/ccs</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102488166</td>\n",
       "      <td>19875</td>\n",
       "      <td>20</td>\n",
       "      <td>41</td>\n",
       "      <td>0.002063</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000553</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>102508040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7358</th>\n",
       "      <td>m64076_210330_204128/60688176/ccs</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>103717856</td>\n",
       "      <td>11074</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103728929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7359</th>\n",
       "      <td>m64076_210219_011735/107479947/ccs</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>103718106</td>\n",
       "      <td>11360</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000616</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103729465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7360</th>\n",
       "      <td>m54329U_220227_232630/122227393/ccs</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>103720334</td>\n",
       "      <td>17839</td>\n",
       "      <td>60</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0.000897</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103738172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7361</th>\n",
       "      <td>m64076_210221_195503/62851886/ccs</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>103720910</td>\n",
       "      <td>10529</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103731438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7362</th>\n",
       "      <td>m64076_210219_011735/8718492/ccs</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>103722334</td>\n",
       "      <td>10562</td>\n",
       "      <td>60</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>8</td>\n",
       "      <td>139</td>\n",
       "      <td>0.013160</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103732895</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3497 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                read_name        chromosome      start  \\\n",
       "3866  m54329U_220313_173323/117639354/ccs  chr15_hap2_hsa14  102481413   \n",
       "3867   m54329U_210404_020346/15336368/ccs  chr15_hap2_hsa14  102487172   \n",
       "3868   m54329U_220313_173323/76744132/ccs  chr15_hap2_hsa14  102487651   \n",
       "3869    m64076_220312_125202/28902388/ccs  chr15_hap2_hsa14  102487840   \n",
       "3870   m64076_220313_224245/157485750/ccs  chr15_hap2_hsa14  102488166   \n",
       "...                                   ...               ...        ...   \n",
       "7358    m64076_210330_204128/60688176/ccs  chr15_hap2_hsa14  103717856   \n",
       "7359   m64076_210219_011735/107479947/ccs  chr15_hap2_hsa14  103718106   \n",
       "7360  m54329U_220227_232630/122227393/ccs  chr15_hap2_hsa14  103720334   \n",
       "7361    m64076_210221_195503/62851886/ccs  chr15_hap2_hsa14  103720910   \n",
       "7362     m64076_210219_011735/8718492/ccs  chr15_hap2_hsa14  103722334   \n",
       "\n",
       "      read_length  mapping_quality  mismatches  mismatch_rate  longindels  \\\n",
       "3866        28540               30           5       0.000175           0   \n",
       "3867        19795               15           0       0.000000           0   \n",
       "3868        28758               60           4       0.000139           0   \n",
       "3869        21075               17           0       0.000000           0   \n",
       "3870        19875               20          41       0.002063           3   \n",
       "...           ...              ...         ...            ...         ...   \n",
       "7358        11074                4           0       0.000000           0   \n",
       "7359        11360                5           7       0.000616           0   \n",
       "7360        17839               60           1       0.000056           2   \n",
       "7361        10529               23           0       0.000000           0   \n",
       "7362        10562               60           4       0.000379           8   \n",
       "\n",
       "      total_indel_length  indel_rate  soft_clipping  hard_clipping        end  \n",
       "3866                   0    0.000000              3              0  102509952  \n",
       "3867                   0    0.000000              0              0  102506966  \n",
       "3868                   0    0.000000              0              0  102516408  \n",
       "3869                   0    0.000000              0              0  102508914  \n",
       "3870                  11    0.000553              0              0  102508040  \n",
       "...                  ...         ...            ...            ...        ...  \n",
       "7358                   0    0.000000              0              0  103728929  \n",
       "7359                   0    0.000000              0              0  103729465  \n",
       "7360                  16    0.000897              0              0  103738172  \n",
       "7361                   0    0.000000              0              0  103731438  \n",
       "7362                 139    0.013160              0              0  103732895  \n",
       "\n",
       "[3497 rows x 13 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "igh = pd.read_csv(f\"../errorStats/{species}/{gene}.txt\", sep=\"\\t\", names=['read_name','chromosome','start','read_length','mapping_quality','mismatches','mismatch_rate','longindels','total_indel_length','indel_rate','soft_clipping','hard_clipping'])\n",
    "igh[\"start\"] = pd.to_numeric(igh[\"start\"])\n",
    "igh[\"read_length\"] = pd.to_numeric(igh[\"read_length\"])\n",
    "igh[\"mapping_quality\"] = pd.to_numeric(igh[\"mapping_quality\"])\n",
    "igh['mismatches'] = igh['mismatches'].astype('int')\n",
    "igh['mismatch_rate'] = igh['mismatch_rate'].astype('float')\n",
    "igh['longindels'] = igh['longindels'].astype('int')\n",
    "igh['total_indel_length'] = igh['total_indel_length'].astype('int')\n",
    "igh['indel_rate'] = igh['indel_rate'].astype('float')\n",
    "igh['soft_clipping'] = igh['soft_clipping'].astype('int')\n",
    "igh['hard_clipping'] = igh['hard_clipping'].astype('int')\n",
    "igh['end'] = igh['start'] + igh['read_length'] - 1\n",
    "chr1 = pri_pileup['Chrom'].value_counts().index[0]\n",
    "igh_pri = igh[(igh['chromosome'] == chr1)]\n",
    "igh_pri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "0cdef789-e992-4d72-ba46-529944a69541",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LocusPos</th>\n",
       "      <th>Locus</th>\n",
       "      <th>GeneType</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Productive</th>\n",
       "      <th>SeqLength</th>\n",
       "      <th>EndPos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20000</td>\n",
       "      <td>IGH</td>\n",
       "      <td>J</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102516796</td>\n",
       "      <td>-</td>\n",
       "      <td>ATTACTACGGTATGGACTTCTGGGACTAAGGGACCATGGTCACCGT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>63</td>\n",
       "      <td>102516858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21021</td>\n",
       "      <td>IGH</td>\n",
       "      <td>J</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102517817</td>\n",
       "      <td>-</td>\n",
       "      <td>ACTACTTTGACTACTGGGGCCAGGGAACCCTGCTCACCGTCTCCTCAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48</td>\n",
       "      <td>102517864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21393</td>\n",
       "      <td>IGH</td>\n",
       "      <td>J</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102518189</td>\n",
       "      <td>-</td>\n",
       "      <td>TGATGCTTTTGATATCTGGGGCCTAGGGGTCGTGGTCACCGTCTCT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50</td>\n",
       "      <td>102518238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22005</td>\n",
       "      <td>IGH</td>\n",
       "      <td>J</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102518801</td>\n",
       "      <td>-</td>\n",
       "      <td>CTACTGGTACTTCGATCTCTGGGGCCGTGGCACCCTGGTCACCGTC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53</td>\n",
       "      <td>102518853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22213</td>\n",
       "      <td>IGH</td>\n",
       "      <td>J</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102519009</td>\n",
       "      <td>-</td>\n",
       "      <td>GCCGAATACTTCAAGCACTGGGGCCAGGGCATCCTGGTCACCGTCT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>52</td>\n",
       "      <td>102519060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>22356</td>\n",
       "      <td>IGH</td>\n",
       "      <td>D</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102519152</td>\n",
       "      <td>+</td>\n",
       "      <td>TCCCCAGTTAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>102519162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>44167</td>\n",
       "      <td>IGH</td>\n",
       "      <td>D</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102540963</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTATAGTGGGAGCTACTAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20</td>\n",
       "      <td>102540982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>44674</td>\n",
       "      <td>IGH</td>\n",
       "      <td>D</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102541470</td>\n",
       "      <td>+</td>\n",
       "      <td>GTAGCCACTGCTATACCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18</td>\n",
       "      <td>102541487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>44674</td>\n",
       "      <td>IGH</td>\n",
       "      <td>D</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102541470</td>\n",
       "      <td>-</td>\n",
       "      <td>GGGTATAGCAGTGGCTAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18</td>\n",
       "      <td>102541487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>51433</td>\n",
       "      <td>IGH</td>\n",
       "      <td>D</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102548229</td>\n",
       "      <td>-</td>\n",
       "      <td>AGGATATTGTAGTGGTGGTAGCTGCTACGCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31</td>\n",
       "      <td>102548259</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   LocusPos Locus GeneType            Contig        Pos Strand  \\\n",
       "0     20000   IGH        J  chr15_hap2_hsa14  102516796      -   \n",
       "1     21021   IGH        J  chr15_hap2_hsa14  102517817      -   \n",
       "2     21393   IGH        J  chr15_hap2_hsa14  102518189      -   \n",
       "3     22005   IGH        J  chr15_hap2_hsa14  102518801      -   \n",
       "4     22213   IGH        J  chr15_hap2_hsa14  102519009      -   \n",
       "5     22356   IGH        D  chr15_hap2_hsa14  102519152      +   \n",
       "6     44167   IGH        D  chr15_hap2_hsa14  102540963      -   \n",
       "7     44674   IGH        D  chr15_hap2_hsa14  102541470      +   \n",
       "8     44674   IGH        D  chr15_hap2_hsa14  102541470      -   \n",
       "9     51433   IGH        D  chr15_hap2_hsa14  102548229      -   \n",
       "\n",
       "                                            Sequence Productive  SeqLength  \\\n",
       "0  ATTACTACGGTATGGACTTCTGGGACTAAGGGACCATGGTCACCGT...        NaN         63   \n",
       "1   ACTACTTTGACTACTGGGGCCAGGGAACCCTGCTCACCGTCTCCTCAG        NaN         48   \n",
       "2  TGATGCTTTTGATATCTGGGGCCTAGGGGTCGTGGTCACCGTCTCT...        NaN         50   \n",
       "3  CTACTGGTACTTCGATCTCTGGGGCCGTGGCACCCTGGTCACCGTC...        NaN         53   \n",
       "4  GCCGAATACTTCAAGCACTGGGGCCAGGGCATCCTGGTCACCGTCT...        NaN         52   \n",
       "5                                        TCCCCAGTTAG        NaN         11   \n",
       "6                               GGTATAGTGGGAGCTACTAC        NaN         20   \n",
       "7                                 GTAGCCACTGCTATACCC        NaN         18   \n",
       "8                                 GGGTATAGCAGTGGCTAC        NaN         18   \n",
       "9                    AGGATATTGTAGTGGTGGTAGCTGCTACGCC        NaN         31   \n",
       "\n",
       "      EndPos  \n",
       "0  102516858  \n",
       "1  102517864  \n",
       "2  102518238  \n",
       "3  102518853  \n",
       "4  102519060  \n",
       "5  102519162  \n",
       "6  102540982  \n",
       "7  102541487  \n",
       "8  102541487  \n",
       "9  102548259  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneFile = f'../primate_igtr/{species}_{AssType}_{gene}.csv'\n",
    "genes = pd.read_csv(geneFile)\n",
    "genes['SeqLength'] = genes['Sequence'].apply(len)\n",
    "genes['EndPos'] = genes['Pos'] + genes['SeqLength'] - 1 \n",
    "genes.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "be384e6e-f1ae-466c-ab96-9eb7d6cd52d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LocusPos</th>\n",
       "      <th>Locus</th>\n",
       "      <th>GeneType</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Productive</th>\n",
       "      <th>SeqLength</th>\n",
       "      <th>EndPos</th>\n",
       "      <th>Average_Coverage</th>\n",
       "      <th>Mismatched_Positions</th>\n",
       "      <th>Matched_Positions</th>\n",
       "      <th>Position_Mismatches</th>\n",
       "      <th>Position_Matches</th>\n",
       "      <th>Percent_Accuracy</th>\n",
       "      <th>Positions_With_At_Least_10x_Coverage</th>\n",
       "      <th>Fully_Spanning_Reads</th>\n",
       "      <th>Fully_Spanning_Reads_100%_Match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20000</td>\n",
       "      <td>IGH</td>\n",
       "      <td>J</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102516796</td>\n",
       "      <td>-</td>\n",
       "      <td>ATTACTACGGTATGGACTTCTGGGACTAAGGGACCATGGTCACCGT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>63</td>\n",
       "      <td>102516858</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>42;42;42;42;42;42;42;42;42;42;42;42;42;42;42;4...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21021</td>\n",
       "      <td>IGH</td>\n",
       "      <td>J</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102517817</td>\n",
       "      <td>-</td>\n",
       "      <td>ACTACTTTGACTACTGGGGCCAGGGAACCCTGCTCACCGTCTCCTCAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48</td>\n",
       "      <td>102517864</td>\n",
       "      <td>0.791667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>38;38;38;38;38;38;38;38;38;38;38;38;38;38;38;3...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21393</td>\n",
       "      <td>IGH</td>\n",
       "      <td>J</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102518189</td>\n",
       "      <td>-</td>\n",
       "      <td>TGATGCTTTTGATATCTGGGGCCTAGGGGTCGTGGTCACCGTCTCT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50</td>\n",
       "      <td>102518238</td>\n",
       "      <td>0.760000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>38;38;38;38;38;38;38;39;39;39;39;39;39;39;39;3...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22005</td>\n",
       "      <td>IGH</td>\n",
       "      <td>J</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102518801</td>\n",
       "      <td>-</td>\n",
       "      <td>CTACTGGTACTTCGATCTCTGGGGCCGTGGCACCCTGGTCACCGTC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53</td>\n",
       "      <td>102518853</td>\n",
       "      <td>0.754717</td>\n",
       "      <td>0.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>1;0;0;0;0;0;0;0;0;0;0;1;0;0;0;0;0;0;0;1;0;0;0;...</td>\n",
       "      <td>39;40;40;40;40;40;40;40;40;40;40;39;40;40;40;4...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22213</td>\n",
       "      <td>IGH</td>\n",
       "      <td>J</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>102519009</td>\n",
       "      <td>-</td>\n",
       "      <td>GCCGAATACTTCAAGCACTGGGGCCAGGGCATCCTGGTCACCGTCT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>52</td>\n",
       "      <td>102519060</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>39;39;39;39;39;39;39;39;39;39;39;39;39;39;39;3...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>1124922</td>\n",
       "      <td>IGH</td>\n",
       "      <td>V</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>103621718</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGGTCACCTTGAAGGAGTCTGGTCCTGCACTGGTGAAACCCACAC...</td>\n",
       "      <td>True</td>\n",
       "      <td>301</td>\n",
       "      <td>103622018</td>\n",
       "      <td>0.156146</td>\n",
       "      <td>0.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>47;47;47;47;47;47;47;47;47;47;47;47;47;47;47;4...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>1135248</td>\n",
       "      <td>IGH</td>\n",
       "      <td>V</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>103632044</td>\n",
       "      <td>-</td>\n",
       "      <td>GAGGTGCAGCTGGTGGAGTCTGGGGGAGGCTTGGTCCAGCCTGGGG...</td>\n",
       "      <td>True</td>\n",
       "      <td>302</td>\n",
       "      <td>103632345</td>\n",
       "      <td>0.112583</td>\n",
       "      <td>0.0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;1;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>38;38;38;38;38;38;38;38;38;38;38;38;38;37;38;3...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>1152871</td>\n",
       "      <td>IGH</td>\n",
       "      <td>V</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>103649667</td>\n",
       "      <td>-</td>\n",
       "      <td>GAGGTGCAGCTGTTGGAGTCAGGGGGAGGCTTGGTACAGCCTGAGG...</td>\n",
       "      <td>True</td>\n",
       "      <td>296</td>\n",
       "      <td>103649962</td>\n",
       "      <td>0.165541</td>\n",
       "      <td>0.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>50;50;50;50;50;50;50;50;50;50;50;50;50;50;50;5...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>1194890</td>\n",
       "      <td>IGH</td>\n",
       "      <td>V</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>103691686</td>\n",
       "      <td>-</td>\n",
       "      <td>GAGGTGCAGCTGGTGCAGTCTGCAGCAGAGGTGAAAAGGCCCGGGG...</td>\n",
       "      <td>True</td>\n",
       "      <td>294</td>\n",
       "      <td>103691979</td>\n",
       "      <td>0.190476</td>\n",
       "      <td>0.0</td>\n",
       "      <td>294.0</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...</td>\n",
       "      <td>56;56;56;56;56;56;56;56;56;56;56;56;56;56;56;5...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>294.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>43.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>1215788</td>\n",
       "      <td>IGH</td>\n",
       "      <td>V</td>\n",
       "      <td>chr15_hap2_hsa14</td>\n",
       "      <td>103712584</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGGTGCAGTTGGTGCAGTCTGGTCATGAGGTGAAGCAGCCTGGGG...</td>\n",
       "      <td>True</td>\n",
       "      <td>296</td>\n",
       "      <td>103712879</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>0.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;1;0;0;...</td>\n",
       "      <td>32;32;32;32;32;32;32;32;32;32;32;32;32;32;32;3...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    LocusPos Locus GeneType            Contig        Pos Strand  \\\n",
       "0      20000   IGH        J  chr15_hap2_hsa14  102516796      -   \n",
       "1      21021   IGH        J  chr15_hap2_hsa14  102517817      -   \n",
       "2      21393   IGH        J  chr15_hap2_hsa14  102518189      -   \n",
       "3      22005   IGH        J  chr15_hap2_hsa14  102518801      -   \n",
       "4      22213   IGH        J  chr15_hap2_hsa14  102519009      -   \n",
       "..       ...   ...      ...               ...        ...    ...   \n",
       "91   1124922   IGH        V  chr15_hap2_hsa14  103621718      -   \n",
       "92   1135248   IGH        V  chr15_hap2_hsa14  103632044      -   \n",
       "93   1152871   IGH        V  chr15_hap2_hsa14  103649667      -   \n",
       "94   1194890   IGH        V  chr15_hap2_hsa14  103691686      -   \n",
       "95   1215788   IGH        V  chr15_hap2_hsa14  103712584      -   \n",
       "\n",
       "                                             Sequence Productive  SeqLength  \\\n",
       "0   ATTACTACGGTATGGACTTCTGGGACTAAGGGACCATGGTCACCGT...        NaN         63   \n",
       "1    ACTACTTTGACTACTGGGGCCAGGGAACCCTGCTCACCGTCTCCTCAG        NaN         48   \n",
       "2   TGATGCTTTTGATATCTGGGGCCTAGGGGTCGTGGTCACCGTCTCT...        NaN         50   \n",
       "3   CTACTGGTACTTCGATCTCTGGGGCCGTGGCACCCTGGTCACCGTC...        NaN         53   \n",
       "4   GCCGAATACTTCAAGCACTGGGGCCAGGGCATCCTGGTCACCGTCT...        NaN         52   \n",
       "..                                                ...        ...        ...   \n",
       "91  CAGGTCACCTTGAAGGAGTCTGGTCCTGCACTGGTGAAACCCACAC...       True        301   \n",
       "92  GAGGTGCAGCTGGTGGAGTCTGGGGGAGGCTTGGTCCAGCCTGGGG...       True        302   \n",
       "93  GAGGTGCAGCTGTTGGAGTCAGGGGGAGGCTTGGTACAGCCTGAGG...       True        296   \n",
       "94  GAGGTGCAGCTGGTGCAGTCTGCAGCAGAGGTGAAAAGGCCCGGGG...       True        294   \n",
       "95  CAGGTGCAGTTGGTGCAGTCTGGTCATGAGGTGAAGCAGCCTGGGG...       True        296   \n",
       "\n",
       "       EndPos  Average_Coverage  Mismatched_Positions  Matched_Positions  \\\n",
       "0   102516858          0.666667                   0.0               63.0   \n",
       "1   102517864          0.791667                   0.0               48.0   \n",
       "2   102518238          0.760000                   0.0               50.0   \n",
       "3   102518853          0.754717                   0.0               53.0   \n",
       "4   102519060          0.750000                   0.0               52.0   \n",
       "..        ...               ...                   ...                ...   \n",
       "91  103622018          0.156146                   0.0              301.0   \n",
       "92  103632345          0.112583                   0.0              302.0   \n",
       "93  103649962          0.165541                   0.0              296.0   \n",
       "94  103691979          0.190476                   0.0              294.0   \n",
       "95  103712879          0.108108                   0.0              296.0   \n",
       "\n",
       "                                  Position_Mismatches  \\\n",
       "0   0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...   \n",
       "1   0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...   \n",
       "2   0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...   \n",
       "3   1;0;0;0;0;0;0;0;0;0;0;1;0;0;0;0;0;0;0;1;0;0;0;...   \n",
       "4   0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...   \n",
       "..                                                ...   \n",
       "91  0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...   \n",
       "92  0;0;0;0;0;0;0;0;0;0;0;0;0;1;0;0;0;0;0;0;0;0;0;...   \n",
       "93  0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...   \n",
       "94  0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...   \n",
       "95  0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;1;0;0;...   \n",
       "\n",
       "                                     Position_Matches  Percent_Accuracy  \\\n",
       "0   42;42;42;42;42;42;42;42;42;42;42;42;42;42;42;4...             100.0   \n",
       "1   38;38;38;38;38;38;38;38;38;38;38;38;38;38;38;3...             100.0   \n",
       "2   38;38;38;38;38;38;38;39;39;39;39;39;39;39;39;3...             100.0   \n",
       "3   39;40;40;40;40;40;40;40;40;40;40;39;40;40;40;4...             100.0   \n",
       "4   39;39;39;39;39;39;39;39;39;39;39;39;39;39;39;3...             100.0   \n",
       "..                                                ...               ...   \n",
       "91  47;47;47;47;47;47;47;47;47;47;47;47;47;47;47;4...             100.0   \n",
       "92  38;38;38;38;38;38;38;38;38;38;38;38;38;37;38;3...             100.0   \n",
       "93  50;50;50;50;50;50;50;50;50;50;50;50;50;50;50;5...             100.0   \n",
       "94  56;56;56;56;56;56;56;56;56;56;56;56;56;56;56;5...             100.0   \n",
       "95  32;32;32;32;32;32;32;32;32;32;32;32;32;32;32;3...             100.0   \n",
       "\n",
       "    Positions_With_At_Least_10x_Coverage  Fully_Spanning_Reads  \\\n",
       "0                                   63.0                  42.0   \n",
       "1                                   48.0                  38.0   \n",
       "2                                   50.0                  38.0   \n",
       "3                                   53.0                  40.0   \n",
       "4                                   52.0                  39.0   \n",
       "..                                   ...                   ...   \n",
       "91                                 301.0                  47.0   \n",
       "92                                 302.0                  34.0   \n",
       "93                                 296.0                  49.0   \n",
       "94                                 294.0                  56.0   \n",
       "95                                 296.0                  32.0   \n",
       "\n",
       "    Fully_Spanning_Reads_100%_Match  \n",
       "0                              29.0  \n",
       "1                              24.0  \n",
       "2                              24.0  \n",
       "3                              26.0  \n",
       "4                              25.0  \n",
       "..                              ...  \n",
       "91                             30.0  \n",
       "92                             24.0  \n",
       "93                             36.0  \n",
       "94                             43.0  \n",
       "95                             19.0  \n",
       "\n",
       "[96 rows x 19 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for index, row in genes.iterrows():\n",
    "    chrom = row['Contig']\n",
    "    start = row['Pos']\n",
    "    length = row['SeqLength']\n",
    "    end = row['EndPos']\n",
    "    locus = row['Locus']\n",
    "    geneType = row['GeneType']\n",
    "    gene_df = pri_pileup[(pri_pileup['Pos'] >= start) & (pri_pileup['Pos'] <= end)]\n",
    "    positions_with_10x = (gene_df['Depth'] >= 10).sum()\n",
    "    gene_df['mismatch_rate'] = 100 - gene_df['PercentCorrect']\n",
    "    mismatchPos = (gene_df['mismatch_rate'] > 20).sum()\n",
    "    matchPos = (gene_df['mismatch_rate'] < 20).sum()\n",
    "    reads_spanning_region = igh_pri[(igh_pri['start'] <= start) & (igh_pri['end'] >= end)].shape[0]\n",
    "    Fully_Spanning_Reads_100 = igh_pri[(igh_pri['start'] <= start) & (igh_pri['end'] >= end) & (igh_pri['mismatches'] == 0)].shape[0]\n",
    "    if length > 0:\n",
    "        Average_Coverage = reads_spanning_region / length\n",
    "        percent_accuracy = (matchPos / length) * 100\n",
    "    else:\n",
    "        Average_Coverage = 0\n",
    "        percent_accuracy = 0\n",
    "    Position_Mismatches = ';'.join(map(str, gene_df['Depth']-gene_df['Correct']))\n",
    "    Position_matches = ';'.join(map(str, gene_df['Correct']))\n",
    "    genes.loc[index, 'Average_Coverage'] = Average_Coverage\n",
    "    genes.loc[index, 'Mismatched_Positions'] = mismatchPos\n",
    "    genes.loc[index, 'Matched_Positions'] = matchPos\n",
    "    genes.loc[index, 'Position_Mismatches'] = Position_Mismatches\n",
    "    genes.loc[index, 'Position_Matches'] = Position_matches\n",
    "    genes.loc[index, 'Percent_Accuracy'] = percent_accuracy\n",
    "    genes.loc[index, 'Positions_With_At_Least_10x_Coverage'] = positions_with_10x\n",
    "    genes.loc[index, 'Fully_Spanning_Reads'] = reads_spanning_region\n",
    "    genes.loc[index, 'Fully_Spanning_Reads_100%_Match'] = Fully_Spanning_Reads_100\n",
    "\n",
    "genes    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "5a5679f8-6c6b-43b4-9848-b33412df2a47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LocusPos</th>\n",
       "      <th>Pos</th>\n",
       "      <th>SeqLength</th>\n",
       "      <th>EndPos</th>\n",
       "      <th>Average_Coverage</th>\n",
       "      <th>Mismatched_Positions</th>\n",
       "      <th>Matched_Positions</th>\n",
       "      <th>Percent_Accuracy</th>\n",
       "      <th>Positions_With_At_Least_10x_Coverage</th>\n",
       "      <th>Fully_Spanning_Reads</th>\n",
       "      <th>Fully_Spanning_Reads_100%_Match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>9.600000e+01</td>\n",
       "      <td>9.600000e+01</td>\n",
       "      <td>96.000000</td>\n",
       "      <td>9.600000e+01</td>\n",
       "      <td>96.000000</td>\n",
       "      <td>96.0</td>\n",
       "      <td>96.000000</td>\n",
       "      <td>96.0</td>\n",
       "      <td>96.000000</td>\n",
       "      <td>96.000000</td>\n",
       "      <td>96.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>4.578251e+05</td>\n",
       "      <td>1.029546e+08</td>\n",
       "      <td>189.697917</td>\n",
       "      <td>1.029548e+08</td>\n",
       "      <td>0.768027</td>\n",
       "      <td>0.0</td>\n",
       "      <td>189.697917</td>\n",
       "      <td>100.0</td>\n",
       "      <td>189.697917</td>\n",
       "      <td>43.208333</td>\n",
       "      <td>26.791667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>3.997137e+05</td>\n",
       "      <td>3.997137e+05</td>\n",
       "      <td>133.755088</td>\n",
       "      <td>3.998215e+05</td>\n",
       "      <td>0.906760</td>\n",
       "      <td>0.0</td>\n",
       "      <td>133.755088</td>\n",
       "      <td>0.0</td>\n",
       "      <td>133.755088</td>\n",
       "      <td>8.750238</td>\n",
       "      <td>6.425142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>2.000000e+04</td>\n",
       "      <td>1.025168e+08</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>1.025169e+08</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>30.000000</td>\n",
       "      <td>15.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>7.221600e+04</td>\n",
       "      <td>1.025690e+08</td>\n",
       "      <td>24.000000</td>\n",
       "      <td>1.025690e+08</td>\n",
       "      <td>0.157142</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.000000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>24.000000</td>\n",
       "      <td>35.750000</td>\n",
       "      <td>21.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>3.426625e+05</td>\n",
       "      <td>1.028395e+08</td>\n",
       "      <td>293.500000</td>\n",
       "      <td>1.028398e+08</td>\n",
       "      <td>0.185176</td>\n",
       "      <td>0.0</td>\n",
       "      <td>293.500000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>293.500000</td>\n",
       "      <td>43.500000</td>\n",
       "      <td>26.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>8.542465e+05</td>\n",
       "      <td>1.033510e+08</td>\n",
       "      <td>296.000000</td>\n",
       "      <td>1.033513e+08</td>\n",
       "      <td>1.395161</td>\n",
       "      <td>0.0</td>\n",
       "      <td>296.000000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>296.000000</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>31.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.215788e+06</td>\n",
       "      <td>1.037126e+08</td>\n",
       "      <td>308.000000</td>\n",
       "      <td>1.037129e+08</td>\n",
       "      <td>4.888889</td>\n",
       "      <td>0.0</td>\n",
       "      <td>308.000000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>308.000000</td>\n",
       "      <td>72.000000</td>\n",
       "      <td>43.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           LocusPos           Pos   SeqLength        EndPos  Average_Coverage  \\\n",
       "count  9.600000e+01  9.600000e+01   96.000000  9.600000e+01         96.000000   \n",
       "mean   4.578251e+05  1.029546e+08  189.697917  1.029548e+08          0.768027   \n",
       "std    3.997137e+05  3.997137e+05  133.755088  3.998215e+05          0.906760   \n",
       "min    2.000000e+04  1.025168e+08    9.000000  1.025169e+08          0.108108   \n",
       "25%    7.221600e+04  1.025690e+08   24.000000  1.025690e+08          0.157142   \n",
       "50%    3.426625e+05  1.028395e+08  293.500000  1.028398e+08          0.185176   \n",
       "75%    8.542465e+05  1.033510e+08  296.000000  1.033513e+08          1.395161   \n",
       "max    1.215788e+06  1.037126e+08  308.000000  1.037129e+08          4.888889   \n",
       "\n",
       "       Mismatched_Positions  Matched_Positions  Percent_Accuracy  \\\n",
       "count                  96.0          96.000000              96.0   \n",
       "mean                    0.0         189.697917             100.0   \n",
       "std                     0.0         133.755088               0.0   \n",
       "min                     0.0           9.000000             100.0   \n",
       "25%                     0.0          24.000000             100.0   \n",
       "50%                     0.0         293.500000             100.0   \n",
       "75%                     0.0         296.000000             100.0   \n",
       "max                     0.0         308.000000             100.0   \n",
       "\n",
       "       Positions_With_At_Least_10x_Coverage  Fully_Spanning_Reads  \\\n",
       "count                             96.000000             96.000000   \n",
       "mean                             189.697917             43.208333   \n",
       "std                              133.755088              8.750238   \n",
       "min                                9.000000             30.000000   \n",
       "25%                               24.000000             35.750000   \n",
       "50%                              293.500000             43.500000   \n",
       "75%                              296.000000             50.000000   \n",
       "max                              308.000000             72.000000   \n",
       "\n",
       "       Fully_Spanning_Reads_100%_Match  \n",
       "count                        96.000000  \n",
       "mean                         26.791667  \n",
       "std                           6.425142  \n",
       "min                          15.000000  \n",
       "25%                          21.000000  \n",
       "50%                          26.000000  \n",
       "75%                          31.000000  \n",
       "max                          43.000000  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "041c17b9-8529-4abc-aadd-512148f07cc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "outFile = f\"{dirStat}/{species}.{AssType}.{gene}.csv\"\n",
    "genes = genes.iloc[:, 1:]\n",
    "genes.to_csv(outFile, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "386b4035-9329-444e-a0c7-2131c7e2b260",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b848246-701c-4a82-9327-8c6dbd8342e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73514b45-3c7b-4424-a6d2-0ace82b636d9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "af764e92-6532-47de-b68e-9e79f8923889",
   "metadata": {},
   "source": [
    "## combined csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "587914ed-f734-4f7a-878b-3667897a570d",
   "metadata": {},
   "outputs": [],
   "source": [
    "species = \"mPonAbe1\"\n",
    "AssType = \"hap2\"\n",
    "dirStat = f\"/home1/zhuyixin/zhuyixin_proj/AssmQuality/errorStats/{species}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "2b1676b4-a47a-4c8f-a075-6fa4ded444c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "csv_files = [f for f in os.listdir(dirStat) if AssType in f and f.endswith('.csv')]\n",
    "order = ['IGH', 'IGK', 'IGL', 'TRA', 'TRB', 'TRG']\n",
    "sort_order = {key: i for i, key in enumerate(order)}\n",
    "sorted_filenames = sorted(csv_files, key=lambda x: sort_order[x.split('.')[2]])\n",
    "dataframes = [pd.read_csv(f\"{dirStat}/{filename}\", sep = \"\\t\") for filename in sorted_filenames]\n",
    "combined_df = pd.concat(dataframes, ignore_index=True)\n",
    "combined_df = combined_df.iloc[:, 0:]\n",
    "outFile = f\"{dirStat}/{species}.{AssType}.combined.csv\"\n",
    "combined_df.to_csv(outFile, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "23efbeba-ed03-49af-85d6-21e464bed3e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Pos</th>\n",
       "      <th>SeqLength</th>\n",
       "      <th>EndPos</th>\n",
       "      <th>Average_Coverage</th>\n",
       "      <th>Mismatched_Positions</th>\n",
       "      <th>Matched_Positions</th>\n",
       "      <th>Percent_Accuracy</th>\n",
       "      <th>Positions_With_At_Least_10x_Coverage</th>\n",
       "      <th>Fully_Spanning_Reads</th>\n",
       "      <th>Fully_Spanning_Reads_100%_Match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>281.000000</td>\n",
       "      <td>2.810000e+02</td>\n",
       "      <td>281.000000</td>\n",
       "      <td>2.810000e+02</td>\n",
       "      <td>281.000000</td>\n",
       "      <td>281.0</td>\n",
       "      <td>281.000000</td>\n",
       "      <td>281.0</td>\n",
       "      <td>281.000000</td>\n",
       "      <td>281.000000</td>\n",
       "      <td>281.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>30.836299</td>\n",
       "      <td>8.006739e+07</td>\n",
       "      <td>250.451957</td>\n",
       "      <td>8.006764e+07</td>\n",
       "      <td>0.388673</td>\n",
       "      <td>0.0</td>\n",
       "      <td>250.451957</td>\n",
       "      <td>100.0</td>\n",
       "      <td>250.451957</td>\n",
       "      <td>47.494662</td>\n",
       "      <td>29.911032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>23.911927</td>\n",
       "      <td>4.805310e+07</td>\n",
       "      <td>94.588009</td>\n",
       "      <td>4.805308e+07</td>\n",
       "      <td>0.602992</td>\n",
       "      <td>0.0</td>\n",
       "      <td>94.588009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>94.588009</td>\n",
       "      <td>9.160521</td>\n",
       "      <td>6.789376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.708122e+07</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>1.708150e+07</td>\n",
       "      <td>0.093333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>28.000000</td>\n",
       "      <td>13.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>12.000000</td>\n",
       "      <td>2.265777e+07</td>\n",
       "      <td>280.000000</td>\n",
       "      <td>2.265806e+07</td>\n",
       "      <td>0.153310</td>\n",
       "      <td>0.0</td>\n",
       "      <td>280.000000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>280.000000</td>\n",
       "      <td>41.000000</td>\n",
       "      <td>26.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>26.000000</td>\n",
       "      <td>1.025608e+08</td>\n",
       "      <td>287.000000</td>\n",
       "      <td>1.025608e+08</td>\n",
       "      <td>0.174377</td>\n",
       "      <td>0.0</td>\n",
       "      <td>287.000000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>287.000000</td>\n",
       "      <td>47.000000</td>\n",
       "      <td>30.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>44.000000</td>\n",
       "      <td>1.035571e+08</td>\n",
       "      <td>296.000000</td>\n",
       "      <td>1.035574e+08</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>0.0</td>\n",
       "      <td>296.000000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>296.000000</td>\n",
       "      <td>54.000000</td>\n",
       "      <td>34.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>95.000000</td>\n",
       "      <td>1.445671e+08</td>\n",
       "      <td>317.000000</td>\n",
       "      <td>1.445674e+08</td>\n",
       "      <td>4.888889</td>\n",
       "      <td>0.0</td>\n",
       "      <td>317.000000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>317.000000</td>\n",
       "      <td>74.000000</td>\n",
       "      <td>51.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0           Pos   SeqLength        EndPos  Average_Coverage  \\\n",
       "count  281.000000  2.810000e+02  281.000000  2.810000e+02        281.000000   \n",
       "mean    30.836299  8.006739e+07  250.451957  8.006764e+07          0.388673   \n",
       "std     23.911927  4.805310e+07   94.588009  4.805308e+07          0.602992   \n",
       "min      0.000000  1.708122e+07    9.000000  1.708150e+07          0.093333   \n",
       "25%     12.000000  2.265777e+07  280.000000  2.265806e+07          0.153310   \n",
       "50%     26.000000  1.025608e+08  287.000000  1.025608e+08          0.174377   \n",
       "75%     44.000000  1.035571e+08  296.000000  1.035574e+08          0.206897   \n",
       "max     95.000000  1.445671e+08  317.000000  1.445674e+08          4.888889   \n",
       "\n",
       "       Mismatched_Positions  Matched_Positions  Percent_Accuracy  \\\n",
       "count                 281.0         281.000000             281.0   \n",
       "mean                    0.0         250.451957             100.0   \n",
       "std                     0.0          94.588009               0.0   \n",
       "min                     0.0           9.000000             100.0   \n",
       "25%                     0.0         280.000000             100.0   \n",
       "50%                     0.0         287.000000             100.0   \n",
       "75%                     0.0         296.000000             100.0   \n",
       "max                     0.0         317.000000             100.0   \n",
       "\n",
       "       Positions_With_At_Least_10x_Coverage  Fully_Spanning_Reads  \\\n",
       "count                            281.000000            281.000000   \n",
       "mean                             250.451957             47.494662   \n",
       "std                               94.588009              9.160521   \n",
       "min                                9.000000             28.000000   \n",
       "25%                              280.000000             41.000000   \n",
       "50%                              287.000000             47.000000   \n",
       "75%                              296.000000             54.000000   \n",
       "max                              317.000000             74.000000   \n",
       "\n",
       "       Fully_Spanning_Reads_100%_Match  \n",
       "count                       281.000000  \n",
       "mean                         29.911032  \n",
       "std                           6.789376  \n",
       "min                          13.000000  \n",
       "25%                          26.000000  \n",
       "50%                          30.000000  \n",
       "75%                          34.000000  \n",
       "max                          51.000000  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f9554867-c2f4-47b3-859f-02f6024386b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "species = \"mPonPyg2\"\n",
    "AssType = \"hap2\"\n",
    "dirStat = f\"/home1/zhuyixin/zhuyixin_proj/AssmQuality/errorStats/{species}\"\n",
    "order = ['IGH', 'IGK', 'IGL', 'TRA', 'TRB', 'TRG']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7fdd8a9e-1993-4c22-98ba-f57eb64a43d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IGH\n",
      "29.557833800636402 0.9999759207310466\n",
      "IGK\n",
      "33.188198020410596 0.9999833518810454\n",
      "IGL\n",
      "30.192151540724858 0.9999460195090056\n",
      "TRA\n",
      "31.91613480894697 0.9999797194387152\n",
      "TRB\n",
      "32.27359826284186 0.9999806614074904\n",
      "TRG\n",
      "28.858019833149694 0.9999790125400073\n"
     ]
    }
   ],
   "source": [
    "if (AssType == \"mat\") or (AssType == \"hap1\"):\n",
    "    atype = \"pri\"\n",
    "else:\n",
    "    atype = \"alt\"\n",
    "for gene in order:\n",
    "    pri_pileup_file = f'../errorStats/{species}/{gene}_{atype}_pileup.txt'\n",
    "    results_pri = {}  # To store results, keyed by (chromosome, position)\n",
    "    with open(pri_pileup_file, 'r') as f:\n",
    "        for line in f:\n",
    "            chrom, pos, ref_base, depth, read_bases, _ = line.split()[:6]\n",
    "            correct = parse_read_bases(read_bases)\n",
    "            indel = count_indels(read_bases)            \n",
    "            if int(depth) > 0:\n",
    "                percent_correct = (correct / int(depth)) * 100\n",
    "            else:\n",
    "                percent_correct = 0\n",
    "             # Store or process results\n",
    "            results_pri[(chrom, pos)] = (correct, percent_correct, depth, indel)\n",
    "    \n",
    "    pri_p = [(*key, *value) for key, value in results_pri.items()]\n",
    "    pri_pileup = pd.DataFrame(pri_p, columns=['Chrom', 'Pos', 'Correct', 'PercentCorrect', \"Depth\", \"Indel\"])\n",
    "    pri_pileup['Pos'] = pd.to_numeric(pri_pileup['Pos'])\n",
    "    pri_pileup['Correct'] = pd.to_numeric(pri_pileup['Correct'])\n",
    "    pri_pileup['PercentCorrect'] = pd.to_numeric(pri_pileup['PercentCorrect'])\n",
    "    pri_pileup['Depth'] = pd.to_numeric(pri_pileup['Depth'])\n",
    "    pri_pileup['Indel'] = pd.to_numeric(pri_pileup['Indel'])\n",
    "    pri_pileup.sort_values(by=['Chrom', 'Pos'], inplace=True)\n",
    "    print(gene)\n",
    "    print(pri_pileup['Depth'].mean(), pri_pileup[pri_pileup['PercentCorrect']> 80].shape[0]/pri_pileup.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbf2a5d0-2c56-4ceb-8a6e-55ba219e6544",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bde1487c-6351-4f70-a3a3-649f3c75ce60",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "assembly",
   "language": "python",
   "name": "assembly"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
